File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/linked_list/recursion/P92_ReverseLinkedListII.py
from typing import Optional

from com.dsa.leetcode.labuladong.linked_list.ListNode import ListNode


class Solution:
    def __init__(self):
        self.successor = None

    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == 1:
            return self.reverseN(head, right)
        head.next = self.reverseBetween(head.next, left - 1, right - 1)
        return head

    def reverseN(self, head, n):
        if n == 1:
            self.successor = head.next
            return head
        newHead = self.reverseN(head.next, n - 1)
        head.next.next = head
        head.next = self.successor
        return newHead

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/linked_list/recursion/P92_ReverseLinkedListII.java
package com.dsa.leetcode.labuladong.linked_list.recursion;

import com.dsa.leetcode.labuladong.linked_list.ListNode;

/**
 * @since 2023/12/5 20:51
 */
public class P92_ReverseLinkedListII {

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode node1 = new ListNode(2);
        ListNode node2 = new ListNode(3);
        ListNode node3 = new ListNode(4);
        ListNode node4 = new ListNode(5);
        ListNode node5 = new ListNode(6);
        ListNode node6 = new ListNode(7);
        head.next = node1;
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node6;

        P92_ReverseLinkedListII p92 = new P92_ReverseLinkedListII();
        ListNode listNode = p92.reverseBetween(head, 3, 5);
        while (listNode != null) {
            System.out.println(listNode.val);
            listNode = listNode.next;
        }
    }

    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (left == 1) {
            return reverseN(head, right);
        }
        head.next = reverseBetween(head.next, left - 1, right - 1);
        return head;
    }

    ListNode successor;
    private ListNode reverseN(ListNode head, int n) {
        if (n == 1) {
            successor = head.next;
            return head;
        }
        ListNode newHead = reverseN(head.next, n - 1);
        head.next.next = head;
        head.next = successor;
        return newHead;
    }

}
